Metadata: Rename core.meta to core.meta_conventions #5224

voetberg opened the pull request:
Renaming `meta` to `meta_conventions` - updates the core and internal api calls to following the naming scheme. For the web api, a legacy endpoint is included with a warning it will be depreciated.

bari12 commented on lib/rucio/api/did.py:
```diff
@@ -15,14 +15,15 @@
 
 from copy import deepcopy
 from typing import TYPE_CHECKING
+from rucio.core import meta_conventions as meta_convention_core
```
Move below to the other `rucio.core` imports

bari12 commented on lib/rucio/client/client.py:
```diff
@@ -40,7 +40,7 @@
 class Client(AccountClient,
              AccountLimitClient,
-             MetaClient,
+             MetConventionsClient,
```
`MetaConventionClient` :-)

bari12 commented on lib/rucio/client/metaconventionsclient.py:
```diff
@@ -23,7 +23,7 @@
 
 
-class MetaClient(BaseClient):
+class MetConventionsClient(BaseClient):
```
Please update the documentation descriptions here as well.

bari12 commented on lib/rucio/core/meta_conventions.py:
```diff
@@ -32,9 +32,9 @@
+def add_key(key: str, key_type: Union[KeyType, str], value_type: Optional[str] = None, value_regexp: Optional[str] = None, *, session: "Session"):
     """
-    Adds a new allowed key.
+    Add an allowed key for DID metadata.
```
I am really wondering if we should rename the models below as well.

Thus from `DIDKey` to `DIDMetaConventionKey` or similar
The question is though, if we should als rename the database tables for that. Probably we should

voetberg replied:
I agree, that makes sense to me. It was difficult to figure out wording for some of the documentation even trying to differentiate a metadata key and a metadata conventions key.

voetberg replied:
Something of note while I'm playing around with the tables - after trying to convert the queries in meta_conventions to sqlalchemy 2.0 (re #5247) I noticed a lot of queries that here break when you do this because of foreign key constraints. (e.g.: `(psycopg2.errors.ForeignKeyViolation) update or delete on table "did_keys" violates foreign key constraint "DID_MAP_KEYS_FK" on table "did_key_map"` is raised for any interaction that involves both of the tables, specifically for updating the value of an existing key). 

Is there an additional statements that need to be added to the sqla2.0 queries, is there something running amuck in the model definitions, or are the existing tests just doing unintended things?

bari12 replied:
I don't think so, this really shouldn't be the case if it worked in sqlalchemy 1.0 syntax. Do you have a specific example?

voetberg replied:
Yes, for example swapping 
```
session.query(models.DIDMetaConventionsConstraints.value).\
                filter_by(key=key).\
                filter_by(value=meta[key]).\
                one()

```

for 
```
            statement = select(
                models.DIDMetaConventionsConstraints.value
            ).where(
                and_(models.DIDMetaConventionsConstraints.value == meta[key], models.DIDMetaConventionsConstraints.key == key)
            )
            session.execute(statement).one()

```

Causes this, as does swapping out the `del_key` functionality: 
```
   session.query(models.DIDMetaConventionsKey).filter(key == key).delete()
```
-> 
```
     statement = delete(
             models.DIDMetaConventionsKey
         ).where(
             models.DIDMetaConventionsKey.key == key
         )
     session.execute(statement)
```

voetberg replied:
Nevermind, figured it out. I hadn't been re-syncing the db tables between running tests so things got confused.

bari12 commented on lib/rucio/core/meta_conventions.py:
```diff
@@ -81,9 +81,9 @@ def add_key(key, key_type, value_type=None, value_regexp=None, *, session: "Sess
+def del_key(key: str, *, session: "Session"):
     """
-    Deletes a key.
+    Delete a key in the DID Metadata table.
```
Delete a key in the DID Metadata Convention table

bari12 commented on lib/rucio/core/meta_conventions.py:
```diff
@@ -94,7 +94,7 @@ def del_key(key, *, session: "Session"):
 def list_keys(*, session: "Session"):
     """
-    Lists all keys.
+    Lists all keys for DID Metadata.
```
Same. Really be careful so people do not confuse the conventions with actual metadata

voetberg committed:
Metadata: Rename core.meta to core.meta_conventions Fix #5224

Changes:
	* Rename core.meta to core.meta_conventions
	* Rename DIDKey -> DIDMetaConventionsKey
	* Rename DIDKeyValue -> DIDMetaConventionsConstraints
	* Update docstrings to reflect naming, more descriptive
	* Typing on *.meta_conventions

voetberg committed:
Metadata: sqla2.0 update for 'meta_conventions'; Fix #5224

bari12 commented on lib/rucio/core/meta_conventions.py:
```diff
@@ -77,46 +77,52 @@ def add_key(key, key_type, value_type=None, value_regexp=None, *, session: "Sess
+    statement = select(models.DIDMetaConventionsKey.key)
+    query = session.execute(statement)
     for row in query:
-        key_list.append(row.key)
```
Use `.scalars()` here. so you don't have to do `row[0]`

bari12 commented on lib/rucio/core/meta_conventions.py:
```diff
@@ -137,54 +143,64 @@ def add_value(key, value, *, session: "Session"):
+    ).where(
+        models.DIDMetaConventionsKey.key == key
+    )
+    query = session.execute(statement).one()
```
Also use `.scalars()` here, the `query[0]`, `query[1]` etc is not very nice :-)
See https://github.com/rucio/rucio/blob/2c5b291f53caebf85259010f11791ed7f497572f/lib/rucio/core/rse.py#L1115

bari12 commented on lib/rucio/core/meta_conventions.py:
```diff
@@ -137,54 +143,64 @@ def add_value(key, value, *, session: "Session"):
     """
     value_list = []
-    query = session.query(models.DIDKeyValueAssociation).filter_by(key=key)
+    statement = select(models.DIDMetaConventionsConstraints.value).where(models.DIDMetaConventionsConstraints.key == key)
```
Use `.scalars()`
